extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/read/HivePartitionReaderFactory.scala

```diff
@@ -0,0 +1,197 @@
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.types._
+import org.apache.spark.util.SerializableConfiguration
+
+import org.apache.kyuubi.spark.connector.hive.HiveReader
+
+case class HivePartitionReaderFactory(
+    sqlConf: SQLConf,
+    broadcastHiveConf: Broadcast[SerializableConfiguration],
+    hiveTable: HiveTable,
+    dataSchema: StructType,
+    readDataSchema: StructType,
+    partitionSchema: StructType,
+    partFileToHivePart: Map[PartitionedFile, HivePartition],
+    pushedFilters: Array[Filter] = Array.empty)
+  extends FilePartitionReaderFactory with Logging {
+
+  private val charset: String =
+    sqlConf.getConfString("hive.exec.default.charset", "utf-8")
+
+  val tableDesc = HiveReader.getTableDec(hiveTable)
+  val nonPartitionReadDataKeys = HiveReader.toAttributes(readDataSchema)
+
+  override def buildReader(partitionedFile: PartitionedFile): PartitionReader[InternalRow] = {
+    throw new UnsupportedOperationException("Cannot use buildReader directly.")
+  }
+
+  override def createReader(partition: InputPartition): PartitionReader[InternalRow] = {
+    assert(partition.isInstanceOf[FilePartition])
+    val filePartition = partition.asInstanceOf[FilePartition]
+    val iter: Iterator[HivePartitionedFileReader[InternalRow]] =
+      filePartition.files.toIterator.map { file =>
+        val bindHivePart = partFileToHivePart.getOrElse(file, null)
+        HivePartitionedFileReader(
+          file,
+          new PartitionReaderWithPartitionValues(
+            HivePartitionedReader(
+              file,
+              buildReaderInternal(file, bindHivePart),
+              tableDesc,
+              broadcastHiveConf,
+              nonPartitionReadDataKeys,
+              bindHivePart,
+              charset),
+            readDataSchema,
+            partitionSchema,
+            file.partitionValues))
+      }
+    new FilePartitionReader[InternalRow](iter)
+  }
+
+  def buildReaderInternal(
+      file: PartitionedFile,
+      bindPartition: HivePartition): PartitionReader[Writable] = {
+    val reader = createPartitionWritableReader(file, bindPartition)
+    val fileReader = new PartitionReader[Writable] {
+      override def next(): Boolean = reader.hasNext
+      override def get(): Writable = reader.next()
+      override def close(): Unit = {}
+    }
+    fileReader
+  }
+
+  private def createPartitionWritableReader[T](
+      file: PartitionedFile,
+      bindPartition: HivePartition): Iterator[Writable] = {
+    // Obtain binding HivePartition from input partitioned file
+    val partDesc =
+      if (bindPartition != null) {
+        Utilities.getPartitionDesc(bindPartition)
+      } else null
+
+    val ifc =
+      if (partDesc == null) {
+        hiveTable.getInputFormatClass
+          .asInstanceOf[java.lang.Class[InputFormat[Writable, Writable]]]
+      } else {
+        partDesc.getInputFileFormatClass
```

Comment:
I have some doubt here
Runtime code will always reach here since the in `HiveFileIndex::filterPartitions` all filtered partitions will tranforms into `BindPartition`, right?
Does the code decision here is to handle the inconsistent format between  partition and table?  Can you explain how does this situation will happen and if we dynamically handle the case here, what strategy should be used in (dynamic/static) partition overwrite.

Reply:
> I have some doubt here
> Runtime code will always reach here since the in `HiveFileIndex::filterPartitions` all filtered partitions will tranforms into `BindPartition`, right?

No, if the table is't a partitioned table, partDesc will be null



> Does the code decision here is to handle the inconsistent format between partition and table? Can you explain how does this situation will happen and if we dynamically handle the case here, what strategy should be used in (dynamic/static) partition overwrite.

Hive table partitions have unique metadata, this may be inconsistent with the metadata of the table, Refs: https://spark.apache.org/docs/latest/sql-ref-syntax-ddl-alter-table.html